### ISSUE TITLE
A way around the Web version's websockets

### ISSUE THREAD

**Issue by silver-volt4** [BenjaminUrquhart/Consigliere#5]
I experimented with WebSocket sniffing today. It's pretty simple to make a prototype to redirect WebSocket traffic elsewhere. I was able to log all exchanged data as hex and save it. Upon closer inspection, I noticed that the format is not really changed (I compared player joining and Unity login requests). So this program could work with them, I wanted to pitch some ideas to you.

One solution could be to have Consigliere run a web server, and then to create a userscript to ping all the data to that server for Consigliere to process it. Another option would be to allow some kind of WS communication exporting, and have Consigliere load that. I tried forging a pcap file (pun intended), but Consigliere didn't seem to like it.

I probably would have coded this myself, but I don't know Java and if I tried learning it I'd probably create a monster instead of a solution. Could any of this be possible for you?

**Comment by BenjaminUrquhart** [comment]
I didn't really expect the format to be different, it's more that there's no cross-platform method of redirecting traffic through Consigliere (there's an older incomplete tool that did that, but it only worked on Windows). I don't even know if the required functions are available through Java.

Is there a way you can share this prototype?

**Comment by silver-volt4** [comment]
```
function bufferToHex(buffer) {
    return [...new Uint8Array(buffer)]
        .map(b => b.toString(16).padStart(2, "0"))
        .join("");
}

function addToLog(x) {
    d = new Date();
    window.log += d.getHours().toString() + ":" + d.getMinutes().toString() + ":" + d.getSeconds().toString() + " " + x + "\n";
}

function consig_out(d) {
    console.warn("outbound:",d);
}

function consig_in(d) {
    console.warn("inbound:",d);
}

(function () {

    var ws = window.WebSocket;

    window.WebSocket = function (a, b) {
        var that = b ? new ws(a, b) : new ws(a);
        that.addEventListener("message", (inD) => consig_in(inD.data));
        window.WSSEND = that.send
        that._send = that.send
        that.send = function (u) {
            consig_out(u);
            that._send(u);
        }
        return that;
    };

    window.WebSocket.prototype = ws.prototype;

}());
```

**Comment by silver-volt4** [comment]
Here's the JS code. Paste it in the dev console before logging in. You can modify the functions to inspect the data.
As for getting it to Consigliere, as I said, I could make a userscript that would save all of the websocket data to a file, and you could implement reading of this file.

**Comment by BenjaminUrquhart** [comment]
So this would be like a tampermoney script? Can a script like that connect to a local server?

**Comment by silver-volt4** [comment]
Yes, that could work. I'd personally prefer saving to a file, but forwarding the packets to a server would work too.
